be1-go/cli/mod.go

```diff
@@ -20,34 +23,61 @@ import (
 
 	"github.com/gorilla/websocket"
 
+	"github.com/fsnotify/fsnotify"
 	"github.com/urfave/cli/v2"
 	"golang.org/x/xerrors"
 )
 
+const (
+	// connectionRetryMaxDelay is the maximum time to wait before retrying to connect to a server
+	connectionRetryMaxDelay = 32 * time.Second
+
+	// connectionRetryInitialDelay is the initial time to wait before retrying to connect to a server
+	connectionRetryInitialDelay = 2 * time.Second
+
+	// connectionRetryRate is the rate at which the time to wait before retrying to connect to a server increases
+	connectionRetryRate = 2
+)
+
+// ServerConfig contains the configuration for the server
+type ServerConfig struct {
+	PublicKey      string   `json:"public-key"`
+	PublicAddress  string   `json:"server-public-address"`
+	PrivateAddress string   `json:"server-listen-address"`
+	ClientPort     int      `json:"client-port"`
+	ServerPort     int      `json:"server-port"`
+	OtherServers   []string `json:"other-servers"`
+}
+
 // Serve parses the CLI arguments and spawns a hub and a websocket server for
 // the server
 func Serve(cliCtx *cli.Context) error {
 	log := popstellar.Logger
 
-	// get command line args which specify public key, addresses, port to use for clients
-	// and servers, remote servers address
-	publicAddress := cliCtx.String("server-public-address")
-	privateAddress := cliCtx.String("server-listen-address")
+	configFilePath := cliCtx.String("config-file")
+	var serverConfig ServerConfig
+	var err error
 
-	clientPort := cliCtx.Int("client-port")
-	serverPort := cliCtx.Int("server-port")
-	if clientPort == serverPort {
-		return xerrors.Errorf("client and server ports must be different")
-	}
-	otherServers := cliCtx.StringSlice("other-servers")
+	startedWithConfigFile := false
 
-	pk := cliCtx.String("public-key")
+	// by default, start using a config file
+	serverConfig, err = startWithConfigFile(configFilePath)
+	if err != nil {
```

Comment:
I personnaly would just fail the start of the server if there are any error while reading the config file, and use flags only if there is no config file set.
But if you think your way is best, you don't need to change it.

Reply:
I think your idea is better indeed. It should fail fast in this case. I pushed a commit 30 minutes ago to address your comments. It shows on the branch but not the PR ... (probably because of github issues). I'll re-request a review when everything is back to normal.